File: LSP/typescript/lsp_naive.ts
abstract class Account {
    abstract deposit(amount: number): void;
    abstract withdraw(amount: number): void;
}

class SavingsAccount extends Account {
    private balance: number;

    constructor() {
        super();
        this.balance = 0;
    }

    deposit(amount: number): void {
        this.balance = this.balance + amount;
        console.log("Savings deposit:", amount);
    }

    withdraw(amount: number): void {
        this.balance = this.balance - amount;
        console.log("Savings withdraw:", amount);
    }
}

class CurrentAccount extends Account {
    private balance: number;

    constructor() {
        super();
        this.balance = 0;
    }

    deposit(amount: number): void {
        this.balance = this.balance + amount;
        console.log("Current deposit:", amount);
    }

    withdraw(amount: number): void {
        this.balance = this.balance - amount;
        console.log("Current withdraw:", amount);
    }
}

class FixedDepositAccount extends Account {
    private balance: number;

    constructor() {
        super();
        this.balance = 0;
    }

    deposit(amount: number): void {
        this.balance = this.balance + amount;
        console.log("FD deposit:", amount);
    }

    withdraw(amount: number): void {
        throw new Error("Withdrawal not allowed in Fixed Deposit Account");
    }
}

class BankClient {
    private accounts: Account[];

    constructor(accounts: Account[]) {
        this.accounts = accounts;
    }

    processTransactions(): void {
        for (const account of this.accounts) {
            account.deposit(20000);

            // type checking, bad design
            if (account instanceof FixedDepositAccount) {
                console.log("Withdrawal not allowed in Fixed Deposit Account");
            } else {
                try {
                    account.withdraw(100);
                } 
                catch (error) {
                    console.log((error as Error).message);
                }
            }
        }
    }
}

const accounts: Account[] = [];

accounts.push(new SavingsAccount());
accounts.push(new CurrentAccount());
accounts.push(new FixedDepositAccount());

const bankClient = new BankClient(accounts);
bankClient.processTransactions();

File: LSP/java/LSPNaive.java
import java.util.ArrayList;

abstract class Account {
    abstract void deposit(int amount);
    abstract void withdraw(int amount);
}

class SavingsAccount extends Account {
    private int balance;

    public SavingsAccount() {
        balance = 0;
    }

    @Override
    public void deposit(int amount) {
        balance += amount;
        System.out.println("Deposited: " + amount + " in Savings Account. New Balance: " + balance);
    }

    @Override
    public void withdraw(int amount) {
        if (balance >= amount) {
            balance -= amount;
            System.out.println("Withdrawn: " + amount + " from Savings Account. New Balance: " + balance);
        } else {
            System.out.println("Insufficient funds in Savings Account!");
        }
    }
}

class CurrentAccount extends Account {
    private int balance;

    public CurrentAccount() {
        balance = 0;
    }

    @Override
    public void deposit(int amount) {
        balance += amount;
        System.out.println("Deposited: " + amount + " in Current Account. New Balance: " + balance);
    }

    @Override
    public void withdraw(int amount) {
        if (balance >= amount) {
            balance -= amount;
            System.out.println("Withdrawn: " + amount + " from Current Account. New Balance: " + balance);
        } else {
            System.out.println("Insufficient funds in Current Account!");
        }
    }
}

class FixedDepositAccount extends Account {
    private int balance;

    public FixedDepositAccount() {
        balance = 0;
    }

    @Override
    public void deposit(int amount) {
        balance += amount;
        System.out.println("Deposited: " + amount + " in Fixed Deposit Account. New Balance: " + balance);
    }

    @Override
    public void withdraw(int amount) {
        throw new UnsupportedOperationException("Withdrawal not allowed in Fixed Deposit Account!");
    }
}

class BankClient {
    private ArrayList<Account> accounts;

    public BankClient(ArrayList<Account> accounts) {
        this.accounts = accounts;
    }

    public void processTransactions() {
        for (Account account : accounts) {
            account.deposit(20000);

            // checking type and skipping withdrawal for FixedDepositAccount
            if (account instanceof FixedDepositAccount) {
                System.out.println("Withdrawal not allowed in Fixed Deposit Account!");
            } else {
                try {
                    account.withdraw(100);
                } catch (UnsupportedOperationException e) {
                    System.out.println("Exception: " + e.getMessage());
                }
            }
        }
    }
}

public class LSPNaive {
    public static void main(String[] args) {
        ArrayList<Account> accounts = new ArrayList<>();
        accounts.add(new SavingsAccount());
        accounts.add(new CurrentAccount());
        accounts.add(new FixedDepositAccount());

        BankClient bankClient = new BankClient(accounts);
        bankClient.processTransactions();
    }
}
